magento/inventory
# Fix for 2936 - ReindexAfterSourceItemsDeletePlugin is more efficient

### Description (*)

Currently, any changes to product stock assignments triggers a full rebuild of the `inventory_stock_x` index.

This is caused by the plugin `InventoryIndexer\Plugin\InventoryApi\ReindexAfterSourceItemsDeletePlugin` chaining it's indexing calls down to sourceIndexer, and then over to stockindexer.

The plugin no longer chains it's way down to the stock indexer and will act on the information available to it. There is a "double-pump" for checking source items, to ensure we do pick up and remove items that no longer belong in a particular stock index

### Fixed Issues (if relevant)

1. magento/inventory#2936: `ReindexAfterSourceItemsDeletePlugin` triggers a full stock reindex for a single source removal

### Manual testing scenarios (*)

1. Have a working magento 2.4-develop and inventory 1.2-develop system working with sample data
1. Export the entirety of "Stock Sources" using the magento export process
1. Create one new stock and an additional source.
1. Duplicate the records in the export to include stock for the new sources
1. Enable database query logging
1. Import the modified "Stock Sources" file exported using "Add/Update behaviour" 
1. Consult the database query logs - individual index updates should be observed and a full reindex should not have occured. All products should have appropriate stock levels available in the indexes and in the admin catalog product index page.
1. Import the modified "Stock Sources" file  with import behaviour "Delete"
1. Consult the database query logs - individual index updates should be observed and a full reindex should not have occured. All products should have no stock levels available in the indexes or on the admin catalog product index page.
1. Import the modified "Stock Sources" file using "Replace behaviour" 
1. Consult the database query logs - individual index updates should be observed and a full reindex should not have occured. All products should have appropriate stock levels available in the indexes and in the admin catalog product index page.

### Questions or comments


I need to run through more test scenarios and do some performance testing before I can consider this "perfect". There will also be other scenarios I have not been able to test;
https://github.com/magento/inventory/issues/2002#issuecomment-459662076

### Contribution checklist (*)
